libfastpm/cosmology.c

```diff
@@ -33,86 +181,181 @@ static double growth(double a, FastPMCosmology * c)
+                                               1,
+                                               1);
+    
+    // assume matter domination
+    double aini = 0.00625;  // FIXME: need to make sure this is less than the starting a. For now using z=159.
+    double yini[4];
+    yini[0] = aini;
+    yini[1] = aini;
+    yini[2] = - 3./7. * aini*aini;
+    yini[3] = 2 * yini[2];
+    
+    int status = gsl_odeiv2_driver_apply(drive, &aini, a, yini);
+    if (status != GSL_SUCCESS) {
+        fastpm_raise(-1, "Growth ODE unsuccesful at a=%g.", a);
+    }
+    
+    gsl_odeiv2_driver_free(drive);
+    
+    ode_soln soln;
+    soln.y0 = yini[0];
+    soln.y1 = yini[1];
+    soln.y2 = yini[2];
+    soln.y3 = yini[3];
+    
+    return soln;
 }
 
-double DLogGrowthFactor2(double a, FastPMCosmology * c) {
-    return 2 * pow(OmegaA(a, c), 6.0/11.);
+void fastpm_growth_info_init(FastPMGrowthInfo * growth_info, double a, FastPMCosmology * c) {
+    growth_info->a = a;
+    growth_info->c = c;
+
+    switch (c->growth_mode) {
+        case FASTPM_GROWTH_MODE_LCDM: {
+            double d1 = solve_growth_int(a, c);
+            double d1_a1 = solve_growth_int(1, c);
+            double Om = Omega_m(a, c);
+
+            growth_info->D1 = d1 / d1_a1;
+            growth_info->f1 = pow(Om, 5./9.);
+            /* FIXME: D2 needs to be checked.
+               Note, the Om(a=1) on the denominator, and the lack of the -3/7, is a fix to ensure D_2=1 today.
+               Old code said the 3/7 was absorbed into dx2, but was it? Or is it just to ensure normalisation to 1
+               today. Need to check this, becasue it could mean ODE mode is incorrect.*/
+            growth_info->D2 = growth_info->D1 * growth_info->D1 * pow(Om/Omega_m(1, c), -1./143.);
+            growth_info->f2 = 2 * pow(Om, 6./11.);
+        break; }
+        case FASTPM_GROWTH_MODE_ODE: {
+            ode_soln soln = growth_ode_solve(a, c);
+            ode_soln soln_a1 = growth_ode_solve(1, c);
+
+            growth_info->D1 = soln.y0 / soln_a1.y0;
+            growth_info->f1 = soln.y1 / soln.y0;    /* f = d log D / d log a. Note soln.y1 is d d1 / d log a */
+            growth_info->D2 = soln.y2 / soln_a1.y2;
+            growth_info->f2 = soln.y3 / soln.y2;
+        break; }
+        default:
+            fastpm_raise(-1, "Please enter a valid growth mode.\n");
+    }
 }
 
-double HubbleEa(double a, FastPMCosmology * c)
-{
-    /* H(a) / H0 */
-    return sqrt(c->OmegaM/(a*a*a)+c->OmegaLambda);
-}
-double DHubbleEaDa(double a, FastPMCosmology * c) {
-    /* d E / d a*/
-    double E = HubbleEa(a, c);
-    return 0.5 / E * (-3 * c->OmegaM / (a * a * a * a));
-}
-double D2HubbleEaDa2(double a, FastPMCosmology * c) {
-    double E = HubbleEa(a, c);
-    double dEda = DHubbleEaDa(a, c);
-    return - dEda * dEda / E + dEda * (-4 / a);
+double DGrowthFactorDa(FastPMGrowthInfo * growth_info) {
+    /* dD/da */
+    /* FIXME: Technically the ODE version should agree with LCDM version
```

Comment:
Let's also add a comment like this to the help string of the growth_mode lua parameter.